### PR TITLE
Initial plotter implementation

### PR DESCRIPTION
**Pull request by twe4ked**
This is fun for sine waves but probably not useful in the general case.

Screen Shot 2020-04-05 at 5 30 33 pm

### CODE REVIEW

**Comment by bjeanes** [comment]
@twe4ked and I were just talking about what this might need to do in order to be useful, and I reckon we have to look at more than just a tick's worth of samples to visualise anything musically meaningful, beyond low-frequency sine waves.

Probably having a ring buffer or some other kind of rolling fixed storage of previous samples and then downsampling/quantizing those samples every interval (say, 10th of a second, whatever that is in ticks) and update the waveform (perhaps with some smoothing) to reflect that.

**Comment by twe4ked** [comment]
That's kinda what @charliesome and I had discussed briefly too, he mentioned [`VecDeque`](https://doc.rust-lang.org/std/collections/struct.VecDeque.html) might be useful. Not sure I've got the chops to pull something like that off. Just pushed this up as is for fun and as a place to discuss how to do it “properly”.

**Comment by bjeanes** [comment]
I might take a stab. Seems fun. But I think taht can be a follow-up. This is cool in its own right.

@charliesome we talked a bit about whether this should have two mono inputs, a stereo input, or a single mono input. On one hand, you can always make multiple plotter modules if you needed multiple visualiations (e.g. using a stereo splitter). On the other hand, it could be important to view R+L on the same graph, ensuring same axes etc, and to note divergence. So, two mono inputs seems like a better choice, because you can choose to only connect one. It could be expanded to be a 4- or 8-channel visualiser for maximum flexibility, for instance.

Another opportunity could be to have a linetype which accepted stereo or mono channels and could operate with either. The graph could for instance show 1 or 2 lines.

**Comment by haileys** [comment]
Yeah, I'd like to implement generics for line types. For now I think stereo is probably the most useful input. For now though, maybe showing a separate plot for both left and right (like audacity) is a good idea:

image

Or it could be configurable too

**Comment by twe4ked** [comment]
Screen Shot 2020-04-05 at 7 05 56 pm

**Comment by bjeanes** [comment]
> For now though, maybe showing a separate plot for both left and right (like audacity) is a good idea:

But you can always spawn two plotters if you want to view different charts. That seems more generically useful. If each channel gets its own chart, I'd argue that a single-plot module that takes a mono channel makes more sense, because a StereoSplitter->2xPlotter is an easy composition. But I like primitives...

File: frontend/src/workspace/components/plotter.rs [hunk 1]
```diff
@@ -38,39 +42,54 @@ impl Component for Plotter {
...
-                        .enumerate()
-                        .map(|(x, y)| (x as f32, *y))
-                        .collect::<Vec<(f32, f32)>>();
-                    chart.draw_series(LineSeries::new(series, &colors[i])).unwrap();
+        if let Some(input) = &self.props.indication.inputs[0] {
+            let (channel_1, channel_2): (Vec<f32>, Vec<f32>) = input.into_iter().enumerate().partition_map(|(i, sample)| {
+                if i % 2 == 0 {
+                    Either::Left(sample)
+                } else {
+                    Either::Right(sample)
```
Comment (bjeanes) [inline frontend/src/workspace/components/plotter.rs:50]: This makes me extra happy if the `Either::Left` is the left channel lol. `Either` for an actual binary option beyond just quasi result types (*cough* haskell) is a nice touch.

**Comment by bjeanes** [comment]
A fun usage of this:

image

Incidentally, I think this demonstrates what I was talking about above:

> But you can always spawn two plotters if you want to view different charts. That seems more generically useful. If each channel gets its own chart, I'd argue that a single-plot module that takes a mono channel makes more sense, because a StereoSplitter->2xPlotter is an easy composition. But I like primitives...

**Pull request merged**